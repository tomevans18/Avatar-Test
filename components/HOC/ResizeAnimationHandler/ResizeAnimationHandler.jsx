import React, { Component } from 'react';
import PropTypes from 'prop-types';

import AnimationHandler from './ResizeAnimationHandler.style';

class ResizeAnimationHandler extends Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node,
      PropTypes.string,
    ]).isRequired,
  };

  state = {
    resizing: false,
  };

  componentDidMount() {
    window.addEventListener('resize', this.handleWindowResize.bind(this));
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.handleWindowResize.bind(this));
  }

  handleWindowResize = () => {
    const { resizing } = this.state;

    if (typeof this.timeout !== 'undefined') clearTimeout(this.timeout);

    if (!resizing) {
      this.setState({
        resizing: true,
      });
    }

    this.timeout = setTimeout(() => {
      this.setState({
        resizing: false,
      });
    }, 200);
  };

  render() {
    const { children } = this.props;
    const { resizing } = this.state;
    return <AnimationHandler resizing={resizing}>{children}</AnimationHandler>;
  }
}

export default ResizeAnimationHandler;
