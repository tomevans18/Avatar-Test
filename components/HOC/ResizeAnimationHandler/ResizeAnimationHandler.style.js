import styled, { css } from 'styled-components';

export default styled.div`
  display: flex;
  flex-direction: column;
  flex: 1;

  ${p =>
    p.resizing &&
    css`
      * {
        transition: none !important;
      }
    `};
`;
