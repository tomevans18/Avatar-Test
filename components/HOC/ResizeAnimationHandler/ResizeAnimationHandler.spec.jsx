import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import ResizeAnimationHandler from './ResizeAnimationHandler';
import ResizeAnimationHandlerStyles from './ResizeAnimationHandler.style';

const mockRemoveEventListener = jest.fn();
global.removeEventListener = mockRemoveEventListener;

describe('Resize Animation Handler', () => {
  it('should render as expected', () => {
    const wrapper = shallow(
      <ResizeAnimationHandler>
        <div />
      </ResizeAnimationHandler>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should update state on window resize', () => {
    jest.useFakeTimers();
    const wrapper = shallow(
      <ResizeAnimationHandler>
        <div />
      </ResizeAnimationHandler>
    );
    expect(wrapper.state().resizing).toBeFalsy();
    global.dispatchEvent(new Event('resize'));
    expect(wrapper.state().resizing).toBeTruthy();
    jest.runAllTimers();
    expect(wrapper.state().resizing).toBeFalsy();
  });

  it('should not set state again and also clear timeout if already running', () => {
    jest.useFakeTimers();
    const wrapper = shallow(
      <ResizeAnimationHandler>
        <div />
      </ResizeAnimationHandler>
    );
    wrapper.setState({
      resizing: true,
    });
    expect(wrapper.state().resizing).toBeTruthy();
    global.dispatchEvent(new Event('resize'));
    expect(clearTimeout).toHaveBeenCalledTimes(2);
    expect(wrapper.state().resizing).toBeTruthy();
  });

  it('should remove event listen on unmount', () => {
    const wrapper = shallow(
      <ResizeAnimationHandler>
        <div />
      </ResizeAnimationHandler>
    );
    wrapper.unmount();
    global.expect(global.removeEventListener).toHaveBeenCalled();
  });

  it('should render styles correctly', () => {
    const wrapper = mount(<ResizeAnimationHandlerStyles />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render styles correctly on resize', () => {
    const wrapper = mount(<ResizeAnimationHandlerStyles resizing />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
