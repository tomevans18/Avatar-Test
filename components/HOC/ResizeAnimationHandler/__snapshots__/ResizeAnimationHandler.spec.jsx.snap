// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resize Animation Handler should render as expected 1`] = `
<ResizeAnimationHandlerstyle
  resizing={false}
>
  <div />
</ResizeAnimationHandlerstyle>
`;

exports[`Resize Animation Handler should render styles correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<ResizeAnimationHandlerstyle>
  <StyledComponent
    forwardedComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "ResizeAnimationHandlerstyle-sc-1y0v65a-0",
          "isStatic": false,
          "lastClassName": "c0",
          "rules": Array [
            "display:flex;flex-direction:column;flex:1;",
            [Function],
            ";",
          ],
        },
        "displayName": "ResizeAnimationHandlerstyle",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "ResizeAnimationHandlerstyle-sc-1y0v65a-0",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
  >
    <div
      className="c0"
    />
  </StyledComponent>
</ResizeAnimationHandlerstyle>
`;

exports[`Resize Animation Handler should render styles correctly on resize 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c0 * {
  -webkit-transition: none !important;
  transition: none !important;
}

<ResizeAnimationHandlerstyle
  resizing={true}
>
  <StyledComponent
    forwardedComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "ResizeAnimationHandlerstyle-sc-1y0v65a-0",
          "isStatic": false,
          "lastClassName": "c0",
          "rules": Array [
            "display:flex;flex-direction:column;flex:1;",
            [Function],
            ";",
          ],
        },
        "displayName": "ResizeAnimationHandlerstyle",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "ResizeAnimationHandlerstyle-sc-1y0v65a-0",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
    resizing={true}
  >
    <div
      className="c0"
    />
  </StyledComponent>
</ResizeAnimationHandlerstyle>
`;
